### ISSUE TITLE
How to save AuditEvent entities in the same transaction?

### ISSUE THREAD

**Issue by voroninp** [dotnet/efcore#35345]
I want to create audit events using `ChangeTracker`.
These events must be saved in the same transaction with the changed (added, modified, deleted) entities.
Current values of the properties must be read after they were generated (including the values generated in db).

I generate audit events in `TransactionCommittingAsync` method of `DbTransactionIntepceptor`. (transaction is started implicitly)
The problem I face is that state of saved entities is not updated at the moment interceptor is called, and when I attach audit events and call `SaveChanges` on `eventData.Context` entities get persisted second time.

Should I inject `DbContextFactory` in my interceptor, create a new instance, enlist it in the same transaction, and then attach audit events to that instance?

**Comment by voroninp** [comment]
If so I'd request an api to create a fresh instance of DbContext from the existing one.

**Comment by voroninp** [comment]
When is it not the case?

**Comment by voroninp** [comment]
Also, isn't it the risk of accepting changes prematurely? If the commit fails, the change tracker will be in a broken state so to say.

**Comment by kirides** [comment]
shouldn't make a big difference, as if a transaction fails anywhere in the middle, you're forced to grab a fresh DbContext anyways, as the state it's currently in may be broken.

This is why you retry outside of EF Code, e.g. in the controller level, and not in the "repository" level, if that makes sense.

**Comment by voroninp** [comment]
True, but this solves the issue of broken context. 
Yet I am curious how can I prevent auditing things which were not saved in the first place.

**Comment by roji** [comment]
@kirides is right here, you're best off placing retrying your entire unit of work - on a fresh DbContext - from the start; this means that any error triggers restarting the previous context and state. This is also exactly who EF's retrying strategies generally work.

**Comment by voroninp** [comment]
@roji , Yes, I got this point. Less issues to solve ;-)
But what about [this](https://github.com/dotnet/efcore/issues/35345#issuecomment-2550290011)?

`SavedChanges` method of `SaveChangesInterceptor` fires after implicitly created transaction has been already committed.
That's why I use `DbTRansactionInterpceptor`.

**Comment by roji** [comment]
Any reason you're using SavedChanges and not SavingChanges, which I think is supposed to run before the transaction is committed?

**Comment by voroninp** [comment]
AFAIK, `SavingChanges` is fired before db generated values are set on properties. Am I mistaken?

**Comment by roji** [comment]
Well, yes - SavingChanges is saved before changes are saved to the database, so obviously database-generated values aren't set at that point (the INSERT hasn't occured yet). But if your audit entities reference the things they audit, they'll have a foreign key which will get set by SavingChanges as usual (exactly like when you save a Blog and its Posts).

If you don't want to do that, you'll need a separate SaveChanges to save your audit entities - you can trigger that from the SavedChanges interceptor (and make sure to not have that run on the second, auditing SaveChanges, e.g. via an async local), or simply by overriding the SaveChanges() method itself.

**Comment by voroninp** [comment]
> you can trigger that from the SavedChanges interceptor

I tried it first, but `eventData.Context.Database.CurrentTransaction` was null, so I assumed commit happens before the `SavedChanges` method was called.

**Comment by roji** [comment]
> so I assumed commit happens before the SavedChanges method was called.

If there's no user-managed transaction, that's probably the case (haven't checked). If you're looking to do things with two separate SaveChanges() - which is what I'm understanding - you'll need to manage transactions on your own, at which point you're in control exactly when the transaction is committed etc.

In any case, is there any additional open question here?

**Comment by voroninp** [comment]
> you'll need to manage transactions on your own

Is it safe to do it inside interceptor's code? Start transaction in `SavingChanges` (if it's not yet started) and `commit it in `SavedChanges`?

**Comment by stevendarby** [comment]
@voroninp 

This issue I raised may be of interest, as the code in the collapsed section shows a rough outline of how I have handled auditing in a SaveChangesAsync override in the past. It does indeed create a second context for the audits (something you suggested in your initial post), which means a transaction can be started within SaveChangesAsync and the audits can be cleared on a retry without clearing the changes in the main context, which we don't know about. In this example it uses a different context type, but it could be the same type, created from a factory. (There's a few things I'd do differently now but think it holds up as an example of one way to do it!)

https://github.com/dotnet/efcore/issues/25946
